tests/lms/models/groups_test.py

```diff
@@ -0,0 +1,172 @@
+# -*- coding: utf-8 -*-
+
+import pytest
+from sqlalchemy.exc import IntegrityError
+
+from lms.models import Group
+
+
+@pytest.mark.usefixtures("noise_groups")
+class TestGroup:
+    def test_get_returns_matching_group(self, db_session):
+        db_session.add(
+            Group(
+                tool_consumer_instance_guid="TEST_GUID",
+                context_id="TEST_CONTEXT",
+                pubid="TEST_PUBID",
+            )
+        )
+
+        group = Group.get(db_session, "TEST_GUID", "TEST_CONTEXT")
+
+        assert group.tool_consumer_instance_guid == "TEST_GUID"
+        assert group.context_id == "TEST_CONTEXT"
+        assert group.pubid == "TEST_PUBID"
+
+    def test_get_returns_None_if_no_matching_group(self, db_session):
+        assert Group.get(db_session, "TEST_GUID", "TEST_CONTEXT") is None
+
+    def test_you_cant_save_a_group_with_no_tool_consumer_instance_guid(
+        self, db_session
+    ):
+        db_session.add(Group(context_id="TEST_CONTEXT", pubid="TEST_PUBID"))
+
+        with pytest.raises(
+            IntegrityError,
```

Comment:
This is handled by the view decorator not trying to add a new group to the DB if one already exists, so the `IntegrityError` is never raised.

There _is_ a race condition though, that if two requests for the same course that doesn't have a group yet are being processed concurrently then it's possible that:

1. Request A checks that there's no group in the DB yet - there isn't
2. Request B checks that there's no group in the DB yet - there isn't
3. Request A tries to save a new group to the DB, succeeds
4. Request B tries to save a new group to the DB - `IntegrityError`.

Request B would trigger Pyramid's error view. If the user reloaded the page (e.g. launched the assignment again) it would pick up the group created by request A and not try to create a duplicate group again.

I don't think we should try to program around this problem in the LMS app though as doing so would be quite awkward and difficult and would have to be re-implemented by every client of our groups API. Even if we were to go to the trouble of handling such `IntegrityError`s in the LMS app, it still would have created two groups for the same course in h, one of which will never be used.

I'm afraid this is (yet another) problem with the tracking of pubids in the LMS app. It requires that group upsert API in h using a unique group ID provided by the LMS app instead of an h-generated pubid. Then the LMS app just deterministically generates a group ID from LTI data (SHA1 of tool_consumer_instance_guid and context_id) and throws it at the group upsert API, and doesn't care whether the group already existed or not, and no problems with concurrency. (And group names would also be updated if the course name in the LMS changes.)

So I think we should plan to improve this by implementing group upsert with provider ID in h when we can.